### ISSUE TITLE
Make pkg/catalog Ticker announcement channel configurable

### ISSUE THREAD

**Issue by draychev** [openservicemesh/osm#1501]
We have a few announcement channels within the Service Mesh Catalog (`pkg/catalog`).  (These are `MeshSpec`, `Certmanager`, `IngressMonitor`, `Namespace`, and `Ticker`)

These are channels to which the respective modules send events when something changes.

One of these channels is the `Ticker` one.  This should not be necessary. It would be nice to have if we want to force update of all connected Envoys at a given interval. This should be made optional.

This Issue proposes we toggle on/off this `Ticker` announcement channel and also configure the tick interfal via a ConfigMap (`osm-config`) setting.  Perhaps a missing value or 0 would indicate this is off.  A present integer value would be a ticker duration value in seconds.

This will toggle:
 - add (or not) the following key/value to the `announcementChannels` map: https://github.com/openservicemesh/osm/blob/872ea6a9c1cb13e8c51555f4426c728fca18684f/pkg/catalog/catalog.go#L60
 - on/off the following goroutine:  https://github.com/openservicemesh/osm/blob/872ea6a9c1cb13e8c51555f4426c728fca18684f/pkg/catalog/catalog.go#L68-L71

Note: Search the repo for mentions of this GitHub issue.

---

Adding a few more TODO's from @shashankram's comment https://github.com/openservicemesh/osm/pull/2016#discussion_r520077203
  - [ ] Document (comment) this feature (the ticker goroutine) - what does it do, why do we need it
  - [ ] Test the ticker - ensure events are propagated (at all and at the given interval)

---

**Scope (please mark with X where applicable)**
- New Functionality          [ ]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ X ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ X ]
- Tests                      [ ]
- CI System                  [ ]
- Project Release            [ ]

**Possible use cases**

**Comment by nitishm** [comment]
@draychev I can take a stab at this one.

**Comment by draychev** [comment]
@nitishm That is so kind of you!
I've tried to put pointers in the issue!  Reach out if you need help!
This is all yours!
Thank you so much!

**Comment by nitishm** [comment]
@draychev  Will this require the addition of a command line arg/flag to OSM CLI as well, since we wish to make this optional?

**Comment by draychev** [comment]
@nitishm I was thinking about making this optional but via the ConfigMap, not the OSM CLI.  Adding it to the `osm-config` ConfigMap would allow us to change this anytime after the OSM Controller has started.  The OSM CLI params, on the other hand, are one-time thing - you set them - they persist for the life of the controller.  But you may want to start the OSM controller without a ticker.  Then you decide that you want to send `refresh` signal to all Envoys once every minute for instance - it would be nice to add something like `refresh_envoys_every_X_minutes:  5` and this enables this ticker.


So I was thinking - we could add another optional key to the `osm-config` ConfigMap.
This PR is a good example of adding a new key to the ConfigMap: https://github.com/openservicemesh/osm/pull/1469/files

If you do not have the key - it'd be defaulted to 0 and I think that should mean that we disable this ticker.
If we see a value greater than 0 - we'd consider that an indicator that we need to enable the ticker and tick with the value `* time.Minute`

So if the ConfigMap had `5` we'd tick once every 5 minutes.

What do you think?

**Comment by nitishm** [comment]
> Then you decide that you want to send `refresh` signal to all Envoys once every minute for instance - it would be nice to add something like `refresh_envoys_every_X_minutes: 5` and this enables this ticker.

Does the controller hot reload when the configuration changes? In that case yes it sounds like a good plan. I have created a PR for the `update` command #1474. Once that goes in we could make this a CLI flag as well. Thoughts?

**Comment by shashankram** [comment]
> > Then you decide that you want to send `refresh` signal to all Envoys once every minute for instance - it would be nice to add something like `refresh_envoys_every_X_minutes: 5` and this enables this ticker.
> 
> @draychev Does the controller hot reload when the configuration changes? In that case yes it sounds like a good plan. I have created a PR for the `update` command #1474. Once that goes in we could make this a CLI flag as well. Thoughts?

I don't see a use case to expose this via a CLI flag yet. For the problem this issue is trying to address, exposing this via a ConfigMap should suffice. I see the ticker config as something that starts up using the default config, with the ConfigMap allowing a way to override it.

**Comment by nitishm** [comment]
@shashankram That makes sense. I will get coding in that case :)

**Comment by nitishm** [comment]
So from what I understand, this configuration may not be work with hot reloads, when the `ConfigMap` changes. 

Here's why,

The expected method to toggle the interval is in [catalog.go#L54](https://github.com/openservicemesh/osm/blob/main/pkg/catalog/catalog.go#L54)

```go
func (mc *MeshCatalog) getAnnouncementChannels() []announcementChannel {
	ticking := make(chan interface{})
	announcementChannels := []announcementChannel{
		{"MeshSpec", mc.meshSpec.GetAnnouncementsChannel()},
		{"CertManager", mc.certManager.GetAnnouncementsChannel()},
		{"IngressMonitor", mc.ingressMonitor.GetAnnouncementsChannel()},
                // NITISH (comment) : Ticker added to announcement channels here
		{"Ticker", ticking},
		{"Namespace", mc.namespaceController.GetAnnouncementsChannel()},
	}
	for _, ep := range mc.endpointsProviders {
		annCh := announcementChannel{ep.GetID(), ep.GetAnnouncementsChannel()}
		announcementChannels = append(announcementChannels, annCh)
	}

        // NITISH (comment) : Ticker created, configured and started here only once.
	go func() {
		ticker := time.NewTicker(updateAtLeastEvery)
		ticking <- ticker.C
	}()
	return announcementChannels
}
```

the `getAnnouncementChannels()` method is invoked only **once** when creating a new `MeshCatalog` instance [catalog.go#L18](https://github.com/openservicemesh/osm/blob/main/pkg/catalog/catalog.go#L18)

```go
// NewMeshCatalog creates a new service catalog
func NewMeshCatalog(namespaceController namespace.Controller, kubeClient kubernetes.Interface, meshSpec smi.MeshSpec, certManager certificate.Manager, ingressMonitor ingress.Monitor, stop <-chan struct{}, cfg configurator.Configurator, endpointsProviders ...endpoint.Provider) *MeshCatalog {
	log.Info().Msg("Create a new Service MeshCatalog.")
	sc := MeshCatalog{
		endpointsProviders: endpointsProviders,
		meshSpec:           meshSpec,
		certManager:        certManager,
		ingressMonitor:     ingressMonitor,
		configurator:       cfg,

		expectedProxies:      make(map[certificate.CommonName]expectedProxy),
		connectedProxies:     make(map[certificate.CommonName]connectedProxy),
		disconnectedProxies:  make(map[certificate.CommonName]disconnectedProxy),
		announcementChannels: set.NewSet(),

		// Kubernetes needed to determine what Services a pod that connects to XDS belongs to.
		// In multicluster scenarios this would be a map of cluster ID to Kubernetes client.
		// The certificate itself would contain the cluster ID making it easy to lookup the client in this map.
		kubeClient: kubeClient,

		namespaceController: namespaceController,
	}

	for _, announcementChannel := range sc.getAnnouncementChannels() {
		sc.announcementChannels.Add(announcementChannel)

	}

	go sc.repeater()
	return &sc
}
```

So any changes made to the `ConfigMap` for the **rebroadcast** interval, after the Service catalog is created, will not apply to the repeater. To activate the new ticker, the osm controller would need a restart.

@draychev @shashankram 
What do you suggest we do? 
If this is a crucial (or much wanted) feature then it might need some code restructuring to permit dynamic ticker restarts. Thoughts on that?

**Comment by nitishm** [comment]
An alternative is to find another home for the `Ticker` announcement channel, somewhere in the `repeater()` method (`MeshCatalog`s `configurator` should return the latest configuration, last read from the `ConfigMap`. Within `repeater()` method we could possibly start another goroutine (create a new `periodicBroadcasts()` method for `MeshCatalog` ??). The `ticker` can be started/stopped/restarted as the config changes?!?!

**Comment by draychev** [comment]
Added a few more items related to the ticker to the description based on https://github.com/openservicemesh/osm/pull/2016#discussion_r520077203

**Comment by eduser25** [comment]
closing, we will discuss if we want to reintroduce it as a separate component if we have concerns over the eventual consistency we follow with k8s APIs.
